File: Algorithms/Island Perimeter.js
// https://leetcode.com/problems/island-perimeter/

/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function(grid) {

  function getPerimeter(y, x, grid) {
    var counts = 0
    
    if (x == 0 || !grid[y][x-1]) counts++
    if (y == 0 || !grid[y-1][x]) counts++
    if (!grid[y+1] || !grid[y+1][x]) counts++
    if (!grid[y][x+1]) counts++
    
    return counts
  }

  var 
    yLen = grid.length, 
    xLen = grid[0].length, 
    counts = 0,
    i,
    j

  for (i = 0; i < yLen; i++) 
    for (j = 0; j < xLen; j++) 
      if (grid[i][j]) 
        counts += getPerimeter(i, j, grid)
    
  return counts
};

// console.log(islandPerimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]))

File: 463-island-perimeter/island_perimeter.go
package solution

//
// solution 1
//

func islandPerimeter(grid [][]int) int {
	var ret int
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == 0 {
				continue
			}
			if i == 0 || grid[i-1][j] == 0 {
				ret++
			}
			if j == len(grid[i])-1 || grid[i][j+1] == 0 {
				ret++
			}
			if i == len(grid)-1 || grid[i+1][j] == 0 {
				ret++
			}
			if j == 0 || grid[i][j-1] == 0 {
				ret++
			}
		}
	}
	return ret
}

//
// solution 2
//

func islandPerimeter2(grid [][]int) int {
	var ret int
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[i]); j++ {
			if grid[i][j] == 0 {
				continue
			}
			ret += 4
			if i != 0 && grid[i-1][j] == 1 {
				ret -= 2
			}
			if j != 0 && grid[i][j-1] == 1 {
				ret -= 2
			}
		}
	}
	return ret
}
